### DISCUSSION TITLE
Proposal: Eliminate stutter and lag spikes via Statistical QoS & KCP Transport Refactoring

### DISCUSSION THREAD

**Discussion by SoongVilda** [rustdesk/rustdesk#14995] [General]
# Proposal: Architectural QoS and Transport Layer Improvements

## Introduction

The current logic behind video and network decision-making in RustDesk is relatively simple. It relies heavily on rigid delay thresholds (e.g., hardcoded 150ms constants) and crude moving averages over tiny sample windows. While functional, this leads to suboptimal experiences under diverse network conditions. For instance, high-latency but high-bandwidth links often suffer unnecessary throttling, while sudden localized congestion creates visible stutter and image degradation. 

I've drafted a phased roadmap for architectural improvements designed to make the remote desktop experience smoother, more resilient, and adaptive. I am opening this issue to share these suggestions with the maintainers for potential integration. 

### Current vs. Proposed Architecture

```mermaid
graph TD
    subgraph Current Architecture
        A1[TestDelay Ping 1s] --> B1[2-Sample Arithmetic Mean]
        B1 --> C1{Absolute Thresholds<br/>e.g., < 150ms}
        C1 -->|Pass| D1[Increase FPS/Ratio]
        C1 -->|Fail| E1[Decrease FPS/Ratio]
    end

    subgraph Proposed Architecture
        A2[TestDelay / OWDV] --> B2[60-Sample Sliding Window]
        B2 --> C2[Calculate P50 / P90 Percentiles]
        C2 --> D2[Subtract Baseline Drift]
        D2 --> E2[Queuing Gradient Slope]
        E2 --> F2{AIMD State Machine}
        H2[Encoder Backpressure Time] --> F2
        I2[Content: Dynamic vs Static] --> F2
        F2 --> G2[Proportional FPS/Ratio Adjustment]
    end
```

The improvements are categorized into three tiers based on implementation difficulty and architectural impact:
- **🟢 Green (Quick Wins):** Simple improvements requiring no architectural changes.
- **🟡 Yellow (Medium Effort):** Requires minor architectural changes (e.g., protobuf schema updates).
- **🔴 Red (Transport Layer):** Harder improvements requiring modifications to the underlying transport crate (`kcp-sys`).

---

## 🟢 Green Tier: Quick Wins (Zero Architectural Changes)

**Scope:** These improvements stay entirely within the existing Rust codebase (primarily `src/server/video_qos.rs` and `src/server/video_service.rs`). They require no protocol changes, no external crate modifications, and no client-side work.

### 1. Statistical Queuing Delay & Percentile Math
* **Problem:** The QoS uses a hard `DELAY_THRESHOLD_150MS` constant and an arithmetic mean over just 2 delay samples (`HISTORY_DELAY_LEN = 2`). High-latency links (e.g., cross-continent) trigger throttling even when bandwidth is fine because absolute RTT is mistaken for congestion. Additionally, OS scheduler hiccups easily skew the tiny 2-sample average.
* **Solution:** Move from absolute delay to a **Queuing Delay Gradient**. By calculating the P50 (median) delay over a larger window (e.g., 60 samples) and subtracting the baseline minimum RTT, we can determine the actual size of the network queue. If the gradient is near zero, the link is clear regardless of the absolute RTT.
* **Details:** Expand `RttCalculator` in `video_qos.rs` with a `.percentile(p: u8)` method to dynamically sort the 60-sample `VecDeque`. Remove `DELAY_THRESHOLD_150MS` completely and replace absolute checks like `if avg_delay < 150` with relative gradient checks like `if gradient < 10` (queue is empty).

### 2. Content-Aware Prioritization
* **Problem:** During congestion, the code applies the same penalty regardless of what is on screen. Dropping bitrate during text/coding work makes text unreadable, while dropping FPS during gaming/video makes motion choppy.
* **Solution:** Use the existing `dynamic_screen` variable to choose *what* to sacrifice during congestion.
* **Details:** Modify the congestion branch in `adjust_ratio()` (`video_qos.rs`). 
  ```rust
  if dynamic_screen {
      // High motion: Sacrifice ratio heavily to preserve FPS for smoothness.
      v = current_ratio * 0.85; 
  } else {
      // Static text: Sacrifice FPS heavily to preserve ratio for text clarity.
      fps = fps.min(10); // Hard clamp FPS downward
      v = current_ratio * 0.98; // Barely touch ratio
  }
  ```

### 3. Hardware Encoder Backpressure
* **Problem:** If the host GPU/CPU is maxed out, frames queue up locally before encoding. This creates massive input lag that network QoS cannot see or fix because the bottleneck is local hardware.
* **Solution:** Measure actual encoding execution time and feed it back into the QoS controller to act as a physical speed limit.
* **Details:** Wrap `handle_one_frame()` in `video_service.rs` with `Instant::now()` to measure execution time. Pass this duration into `check_qos()`. In `video_qos.rs`, compute `sustainable_fps = 1.0 / median_encode_time` and clamp the network target FPS so we never ask the hardware for more frames than it can physically encode.

### 4. Fix `response_delayed` Cliff Edge
* **Problem:** When a ping takes >2000ms, `response_delayed` is true. The code feeds the raw elapsed time (e.g., 5000ms) into the tiny 2-sample delay history, poisoning the moving average. Worse, FPS is instantly hard-clamped to `MIN_FPS + 1` and stays there until the next successful ping, causing a massive visual stutter (a "cliff edge" drop).
* **Solution:** Apply a multiplicative decrease (e.g., `fps = fps / 2`) instead of clamping to 2 FPS. Furthermore, cap the injected delay on timeout to a sane maximum (e.g., 1000ms) so it doesn't permanently ruin the statistical averages.

### 5. Ratio Recovery Hysteresis
* **Problem:** When network delay oscillates exactly around a threshold boundary (e.g., 148ms ↔ 152ms), the ratio ping-pongs between increasing and decreasing every cycle, causing visible flickering in image quality.
* **Solution:** Introduce hysteresis and a cooldown counter. Add `last_ratio_direction` and `ratio_hold_cycles` to the `VideoQoS` struct. After a decrease, force the system to wait a few cycles before allowing an increase, and require crossing a wider threshold band to reverse direction.

---

## 🟡 Yellow Tier: Medium Effort (Minor Architectural Changes)

**Scope:** These improvements require protobuf schema changes and/or client-side awareness.

### 1. Per-Frame Telemetry (OWDV) & Clock Drift Regression
* **Problem:** Current delay relies on round-trip `TestDelay` pings. This hides asymmetric congestion (e.g., client upload is saturated, but video download is fine). Long sessions also suffer from quartz oscillator clock drift, making delay appear to slowly increase over time.
* **Solution:** Move to One-Way Delay Variation (OWDV). 
* **Details:** Add an `int64 timestamp` field to `VideoFrame` in `message.proto`. Inject `hbb_common::get_time()` into frames inside `handle_one_frame()`. The client must read this timestamp, compute the delta, and report it back to the server. Add a least-squares linear regression calculation to `RttCalculator` to filter out hardware clock drift.

### 2. Predictive Delay Modeling (AIMD / Velocity)
* **Problem:** Current FPS adjustments are purely reactive. By the time a delay threshold is crossed, the router buffer is already full and packets are dropping.
* **Solution:** Use an AIMD (Additive Increase, Multiplicative Decrease) state machine driven by the *velocity* (slope) of the delay queue.
* **Details:** Calculate the trend line (slope) of the last N P50 samples. If the slope is positive and steep (`velocity > 2.0`), it means the queue is filling fast decrease FPS multiplicatively *before* it overflows.

---

## 🔴 Red Tier: Transport Layer (Hard Effort)

**Scope:** These changes require modifications to the external `kcp-sys` transport wrapper and deep integration.

```mermaid
sequenceDiagram
    participant App as Video Capture Loop
    participant QoS as VideoQoS
    participant KCP as KCP Transport
    participant UDP as UDP Socket
    
    App->>QoS: Update RTT & Encoder Timings
    App->>KCP: Encode & Send Frame
    KCP->>KCP: Check Socket Congestion (Waitsnd > 2 * Sendwnd)
    alt Socket Dead / Overloaded
        KCP-->>App: Signal Transport Collapse
        App->>QoS: Emergency Drop to Min FPS
    else Packet Loss Detected
        QoS->>QoS: Is Queuing Gradient Rising?
        alt Gradient Rising (True Congestion)
            QoS->>App: Throttle Bitrate
        else Gradient Flat (Spurious Radio Interference)
            KCP->>UDP: Ignore loss, maintain speed
            UDP->>UDP: Inject XOR FEC Parity Packet
        end
    end
```

### 1. Bivariate Congestion Classification
* **Problem:** Standard algorithms assume all packet loss = congestion. On WiFi, packet loss happens due to interference without actual bandwidth exhaustion. Throttling the bitrate in response ruins video quality when the pipe is actually open.
* **Solution:** Correlate transport-layer packet loss with the queuing delay gradient. High loss + rising gradient = genuine congestion (throttle). High loss + flat gradient = spurious radio interference (skip throttle).
* **Details:** Fork `kcp-sys` to expose the raw `ikcpcb` internal statistics (e.g., `cwnd`, `xmit`, `snd_wnd`). Plumb these stats through `KcpStream` up to `VideoQoS`.

### 2. Adaptive FEC & Packet Pacing
* **Problem:** Bursting full keyframes overwhelms routers, causing self-inflicted loss. Furthermore, when KCP loses a packet, ARQ retransmission takes a full round-trip to recover, causing massive input lag.
* **Solution:** Route outbound KCP packets through a leaky bucket pacer. When spurious loss is detected (via Phase 1 above), inject lightweight XOR parity FEC packets directly over the UDP socket. This allows the receiver to reconstruct missing data instantly without waiting for ARQ retransmission.

### 3. Transport Layer Bridging
* **Problem:** Application-layer QoS operates on a ~1-second feedback loop. If the KCP transport collapses immediately (socket buffer exhaustion), the video encoder keeps producing frames and dumping them into a dead socket for seconds before QoS notices.
* **Solution:** Bridge transport health directly to the encoder loop. Expose KCP's `waitsnd()` vs `sendwnd()` ratio. If `waitsnd > 2 * sendwnd`, signal an atomic flag that immediately forces the video capture loop to drop FPS to the minimum until the socket drains.

---

## Conclusion

My initial intention was to implement the Green Tier improvements and submit a Pull Request myself. However, I found the RustDesk build environment to be highly fragile, particularly on Arch-based systems (CachyOS) and within Docker. The reliance on pre-built binary blobs (like `libsciter-gtk.so`) and strict clang/bindgen version requirements resulted in immediate heap corruption aborts (`corrupted size vs. prev_size`) on my system before the application could even launch.

Because I cannot reliably build and test the client GUI locally to verify my work, I decided to compile my architectural research into this issue instead. I kindly request that the maintainers consider these improvements. 

Additionally, please consider providing more robust, modernized build instructions or containerized toolchains that don't rely on brittle dynamic linkage. Lowering the build friction would make it much easier for community members like myself to contribute code directly.

Thank you for your time and for the incredible work on RustDesk!
